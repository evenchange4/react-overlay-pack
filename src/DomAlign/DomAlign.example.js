// @flow
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import DomAlign from '.';

class Component extends React.Component<
  {
    config: Object,
  },
  {
    target?: React.ElementRef<any>,
  },
> {
  state = { target: undefined };
  onRef = (target: React.ElementRef<any>) => this.setState({ target });
  render() {
    const { config } = this.props;
    return (
      <React.Fragment>
        <div
          ref={this.onRef}
          style={{ backgroundColor: 'antiquewhite', height: 100, width: 100 }}
        >
          Target node
        </div>

        <DomAlign config={config} target={this.state.target}>
          <div style={{ backgroundColor: 'aliceblue', height: 50, width: 100 }}>
            Align node
          </div>
        </DomAlign>
      </React.Fragment>
    );
  }
}

storiesOf('DomAlign', module)
  .add(
    'API',
    withInfo({
      text: `
      This is a dom-align integration component.

      API Reference: https://github.com/yiminghe/dom-align
    `,
      inline: true,
      propTables: [DomAlign],
    })(() => (
      <Component
        config={{
          points: ['tl', 'bl'], // bottom-right
          offset: [10, 10],
        }}
      />
    )),
  )
  .add(
    'Center',
    withInfo({
      text: 'TopCenter <~ BottomLeft',
      inline: true,
      propTables: [DomAlign],
    })(() => (
      <Component
        config={{
          points: ['bl', 'tc'],
          offset: [5, 0],
        }}
      />
    )),
  );
