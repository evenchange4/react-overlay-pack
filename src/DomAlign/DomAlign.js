/* global window */
// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import domAlign from 'dom-align';
import rafThrottle from 'raf-throttle';

export type Props = {
  children: React.Node,
  config: Object, // docs: https://github.com/yiminghe/dom-align#config-object-details
  target: React.ElementRef<*>,
  resize?: boolean,
};

class DomAlign extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
    config: PropTypes.shape({
      points: PropTypes.array,
      offset: PropTypes.array,
    }),
    target: PropTypes.object, // eslint-disable-line
    resize: PropTypes.bool,
  };

  static defaultProps = {
    target: undefined,
    resize: false,
  };

  constructor(props: Props) {
    super(props);

    this.source = React.createRef();
    this.align = rafThrottle(() => {
      const { target, config } = this.props;
      const { source } = this;

      if (target && target.current && source && source.current) {
        // Note: Wait for two react instance ready.
        domAlign(source.current, target.current, config);
      }
    });
  }

  componentDidMount() {
    const { resize } = this.props;
    this.align();
    if (resize) {
      window.addEventListener('resize', this.align);
      window.addEventListener('scroll', this.align);
      window.addEventListener('wheel', this.align);
    }
  }

  componentDidUpdate() {
    // TODO: make it async. there is a problem of overlay in dialog case.
    this.timeoutId = setTimeout(() => this.align());
  }

  componentWillUnmount() {
    const { resize } = this.props;
    if (resize) {
      window.removeEventListener('resize', this.align);
      window.removeEventListener('scroll', this.align);
      window.removeEventListener('wheel', this.align);
    }
    if (this.align && this.align.cancel) this.align.cancel();
    if (this.timeoutId) clearTimeout(this.timeoutId);
  }

  timeoutId: ?TimeoutID;

  source: { current: null | React.ElementRef<any> };

  align: () => void;

  render() {
    const { children } = this.props;
    const { source } = this;

    return React.cloneElement(React.Children.only(children), {
      ref: source,
    });
  }
}

export default DomAlign;
