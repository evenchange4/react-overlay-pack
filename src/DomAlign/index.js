// @flow
export { default } from './DomAlign';
