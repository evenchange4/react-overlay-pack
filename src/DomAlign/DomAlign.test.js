/* global window */
// @flow

import * as React from 'react';
import { mount } from 'enzyme';
import DomAlign from '../DomAlign';

jest.mock('raf-throttle', () => a => {
  const fn = a;
  fn.cancel = () => {};
  return fn;
});

it('should add a "scroll & resize" EventListener of window', () => {
  // ref: https://github.com/airbnb/enzyme/issues/426
  const windowEvents = {};
  (window: any).addEventListener = jest.fn((event, cb) => {
    windowEvents[event] = cb;
  });
  (window: any).removeEventListener = jest.fn();

  // Before mount
  expect(windowEvents).toEqual({});
  mount(
    <DomAlign config={{}} target={{}} resize>
      <div style={{ backgroundColor: 'aliceblue', height: 50, width: 100 }}>
        Align node
      </div>
    </DomAlign>,
  );

  // After mount
  expect(windowEvents).toMatchSnapshot();
});
