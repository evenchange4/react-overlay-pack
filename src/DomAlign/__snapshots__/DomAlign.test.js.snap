// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a "scroll & resize" EventListener of window 1`] = `
Object {
  "error": [Function],
  "resize": [Function],
  "scroll": [Function],
  "wheel": [Function],
}
`;
